Record extractions to database

robintw opened the pull request:
## 🧰 Issue
Fixes #263 

## 🚀 Overview: 
Adds ability to record extractions to the database. Now the options are to not record extractions at all, to record them just to the highlighted HTML file, or to record them to the highlighted HTML file and the database. 

## 🤔 Reason: 
More traceability for the user

## 🔨Work carried out:

- [x] Update schema of extractions table
- [x] Update HighlightedFile, Token and Line classes to be able to record data into dicts
- [x] Update DatafileMixin to be able to flush extractions to a dict, and then commit them to the database
- [x] Add tests
- [x] Tests pass

## Confirmations

- [x] I have chosen reviewers for my PR.
- [x] I have chosen an appropriate label for the PR, adding `interactive_review` if reviewers will need to see UI
- [x] I have extended/updated the documentation in `\docs` folder
- [x] Any database content changes (Create, Edit, Delete) are recorded in the Log/Changes tables
- [x] Any database schema changes are implemented via `alembic revision` [transitions](https://pepys-import.readthedocs.io/en/latest/database_migration.html#how-to-use-it-for-developers)
- [x] I have completed the mandatory sections of this document.
- [x] I have deleted any unused sections.

robintw committed:
Update alembic config to fix error with connection commit, and ignore more spatialite tables

robintw committed:
Update schema for Extractions table, plus migrations

robintw committed:
Update migrations for final structure of Extractions table

robintw committed:
Fix bug in token start/end character handling, where the character to be stripped was stripped from the text but not taken off the start/end character.

robintw committed:
Implementation of recording to database functionality - including:

- Changes to HighlightedFile and Datafile, so they keep references to each other
- New instance variables in Datafile to keep pending tokens and a dict from measurement objects to token lists
- Methods to flush pending list into dict, and to submit all extractions
- Changes to Token and Line to add dict of values to datafile lists, ready for later write to database

Note: There is currently no option to have just highlighting but not recording of extractions into the database

robintw committed:
Add support for exporting to database to XML parser

robintw committed:
Add relevant flush calls to all importers

robintw committed:
Add utils module (oops)

robintw committed:
Fix postgres migration to make optional fields optional

robintw committed:
Fix column name to match database

robintw committed:
Fix snapshot merging now we're using the Extractions table - it needs merging at the end, and only copying across entries with datafile ids that have been added

robintw committed:
Add functionality to switch recording to be either nothing, highlighted html, or database

robintw committed:
Record extractions to database by default for Nisida importer

robintw committed:
Fix a load of tests

robintw committed:
Fix test for importer with highlighting turned off

robintw committed:
Fix hanging of test where it was generating large numbers of measurement-objects/tokens to store

robintw committed:
Merge branch 'develop' of github.com:debrief/pepys-import into record-extractions-to-database

robintw committed:
Update docs

robintw committed:
Slightly increase benchmark time threshold now we're doing a little more in the record functions

robintw committed:
Add/update tests

robintw committed:
Merge branch 'develop' of github.com:debrief/pepys-import into record-extractions-to-database

robintw committed:
Add tests for db extraction and fix a few bugs found in the process

robintw committed:
Add to merging test to check extractions copy across properly and don't duplicate

robintw committed:
Reorder migrations, as Update Dash Stored Procedures again will have been applied for the users, but Alter Extractions Table won't have been yet - and they were
previously in the opposite order to that, and therefore the migration may not have been applied properly.

robintw committed:
Update comment

robintw committed:
Add __init__ code for Postgres too

robintw committed:
Add debugging outputs for fixing Windows CI errors

robintw committed:
Change how local validators are imported, so that tests that do imports that run after the config tests don't fail.

Any import running as part of a test that ran _after_ the config file tests was failing because of 'left-over' config from the config file tests
that had set up validators that deliberately failed. This changes how the validators are imported so that they pick up the new config after the config
tests have run. This _could_ cause a slowdown in some situations, but I don't believe local validators are even used by the client yet, so I don't think
it'll be a problem.

robintw committed:
Remove un-needed test

robintw committed:
Remove debugging prints

robintw committed:
Fix latest revisions and upgrade test

robintw committed:
Fix test ordering bug in a far nicer way - now we don't reload validators for each file, we do it once, but also have a function we can call in a test to do it manually

robintw commented:
This is now ready @IanMayo, and all Codacy warnings can be ignored (three of them are for tests using the old unittest style, which require tests to be methods of a class, and one of them is for a parameter which has to be passed as part of patching with pytest).

IanMayo committed:
Merge branch 'develop' into record-extractions-to-database

IanMayo commented on pepys_import/file/importer.py:
```diff
@@ -22,13 +22,20 @@ def __init__(self, name, validation_level, short_name, datafile_type, default_pr
-        self.do_recording = True
+        # By default all importers will record extractions to a highlighted
+        # HTML file, but not record them to the database
+        self.highlighting_level = "html"
```
@robintw - maybe I'm a crusty old Java programmer who is far from `pythonic`, but...

I would always declare these strings as constants, and refer to the constant in the code (to avoid [magic strings](https://deviq.com/antipatterns/magic-strings).  IMHO that avoids typos, and easily lets us find (for example) all of the places where a particular highlighting mode is used.

robintw replied:
I thought you might say that! In python code people tend to use strings rather than constants for these sorts of things, but I can switch to using constants if you'd prefer - I'm happy either way.

IanMayo replied:
Do you have a position on this as we move forward @mew-nsc ? Ours will be the new team-culture from the end of the week ;-)

mew-nsc replied:
"Named constant?" or "Would this be better as an enum?" are some of my most common code review comments... It looks like an enum might make sense here, but I'm conscious that the Python convention is to use strings - I'm guessing we should probably stick with the language conventions?

robintw replied:
We do use named constants elsewhere (eg. for validation levels - which is a similar setting that is configured for an importer) - so we could use them here.
I'm wondering whether switching to use constants for this PR might be a good task for @mew-nsc, as it would give him a good excuse to look through all the code in a reasonable amount of detail and see how it all fits together. I'm happy to do it myself, but just wondered if that might be useful.

mew-nsc replied:
Agree that it would be useful to give me a better understanding of the extractions. Will take a look.

mew-nsc replied:
Pushed the change (went for an enum based approach). Please shout if you have any concerns

IanMayo commented:
@robintw - should we be able to view `extractions` in the `admin-maintenance` page?  Since this data isn't included in any business processes (other than recording) it may be useful to provide (slightly) easier access to it - including for my checking the PR when I only have access to codespaces 🤣

robintw committed:
Allow Extractions table to be viewed in GUI

robintw committed:
Merge branch 'record-extractions-to-database' of github.com:debrief/pepys-import into record-extractions-to-database

robintw commented:
@IanMayo You can now! I keep forgetting that we have a hardcoded list of metadata tables that we can view in the GUI, to stop the user viewing tables that aren't used yet. I've added it to that list, and fixed a couple of things so it displays properly.

IanMayo committed:
Merge branch 'develop' into record-extractions-to-database

mew-nsc committed:
Use an enum instead of bare strings for Highlighter levels

robintw commented:
Looks good @mew-nsc, thanks

IanMayo committed:
Merge branch 'develop' into record-extractions-to-database

IanMayo commented:
Hello @robintw  & @mew-nsc 

GH is reporting a conflict, though it seems relatively benign.  Ok if I fix the text, and merge this PR?

robintw commented:
Yeah that's fine @IanMayo

IanMayo committed:
Merge branch 'develop' into record-extractions-to-database

IanMayo reviewed (dismissed):
Tested by modifying REP importer, and viewing results data in `pepys-admin`.

IanMayo committed:
Try to improve formatting, to keep black happy

IanMayo committed:
Another attempt to please Black

IanMayo commented:
It looks like `black` wants some newlines.  Have tried to fix.

![image](https://user-images.githubusercontent.com/1108513/135111541-b89979fa-7511-4ae6-aca1-2a8cc0d8c42b.png)

IanMayo commented:
Closing & re-opening to try to wake up tests.

robintw committed:
Removed unneeded import

robintw committed:
Fix isort issues

IanMayo reviewed (approved):
Tested & working.
